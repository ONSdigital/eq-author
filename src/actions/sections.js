export const ADD_SECTION = 'ADD_SECTION';
export const ADD_QUESTION = 'ADD_QUESTION';
export const ADD_ANSWER = 'ADD_ANSWER';

export const addSection = name => {
  return {
    type: ADD_SECTION,
    payload: {
      name: name,
    },
  };
};

export const addQuestion = (name, sectionID) => {
  return {
    type: ADD_QUESTION,
    payload: {
      name: name,
      sectionID: sectionID,
    },
  };
};

export const addAnswer = (name, questionID, sectionID) => {
  return {
    type: ADD_ANSWER,
    payload: {
      name: name,
      questionID: questionID,
      sectionID: sectionID,
    },
  };
};
