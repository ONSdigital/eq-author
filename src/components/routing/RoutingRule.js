import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { TransitionGroup } from "react-transition-group";

import Transition from "components/routing/Transition";
import Button from "components/Button";
import IconText from "components/IconText";
import IconRoute from "./icon-route.svg?inline";
import RoutingRuleDestinationSelector from "./RoutingRuleDestinationSelector";
import TextButton from "components/TextButton";
import RoutingCondition from "components/routing/RoutingCondition";
import { get } from "lodash";
import { Grid, Column } from "components/Grid";
import { RADIO } from "constants/answer-types";
import routingRuleFragment from "graphql/fragments/routing-rule.graphql";
import { colors } from "constants/theme";
import { Select as BaseSelect } from "components/Forms";

const Box = styled.div`
  position: relative;
`;

const Padding = styled.div`
  padding: 1em 0;
`;

const Buttons = styled.div`
  display: flex;
  justify-content: flex-end;
  padding: 1em;
`;

const Title = styled.h2`
  position: absolute;
  margin: 0;
  top: 1.5em;
  left: 1.4em;
  letter-spacing: 0.05em;
  font-size: 0.9em;
  font-weight: bold;
  text-transform: uppercase;
`;

const CenteringColumn = styled(Column)`
  display: flex;
  justify-content: center;
  padding: 0.25em 0;
  margin-bottom: 0.5em;
`;

const ConditionSelectorHeader = styled.div`
  background: #e4e8eb;
  border-top: 1px solid ${colors.secondary};
  padding: 0.5em 2em;
  font-weight: bold;
  color: ${colors.secondary};
`;

const Select = styled(BaseSelect)`
  display: inline-block;
  width: auto;
  margin: 0 0.5em;
  padding: 0.3em 2em 0.3em 0.5em;
`;

const ConditionSelect = () => {
  return (
    <Select onChange={() => {}}>
      <option value="all">All of</option>
      <option value="any">Any of</option>
      <option value="none">None of</option>
    </Select>
  );
};

const RoutingRule = ({
  rule,
  onDeleteRule,
  onThenChange,
  onAddRoutingCondition,
  onToggleConditionOption,
  onUpdateRoutingCondition,
  onDeleteRoutingCondition,
  title,
  destinations,
  pagesAvailableForRouting,
  className,
  ...otherProps
}) => {
  const { conditions, id, goto } = rule;
  const existingRadioConditions = {};

  const handleDeleteClick = () => onDeleteRule(rule);
  const handleAddClick = () => onAddRoutingCondition(rule);
  const handleThenChange = value => onThenChange({ id: id, goto: value });

  return (
    <div className={className} data-test="routing-rule">
      <ConditionSelectorHeader>
        Match <ConditionSelect /> the following rules:
      </ConditionSelectorHeader>
      <div>
        {title && <Title>{title}</Title>}

        <Padding>
          <TransitionGroup>
            {conditions.map((condition, index) => {
              const { answer } = condition;
              const canAddAndCondition = !existingRadioConditions[
                get(answer, "id")
              ];

              const component = (
                <Transition key={condition.id}>
                  <div>
                    <RoutingCondition
                      condition={condition}
                      label={index > 0 ? "AND" : "IF"}
                      ruleId={id}
                      sections={pagesAvailableForRouting}
                      onRemove={
                        conditions.length > 1 ? onDeleteRoutingCondition : null
                      }
                      onAdd={handleAddClick}
                      onPageChange={onUpdateRoutingCondition}
                      onToggleOption={onToggleConditionOption}
                      canAddAndCondition={canAddAndCondition}
                      {...otherProps}
                    />
                  </div>
                </Transition>
              );

              if (answer && answer.type === RADIO) {
                existingRadioConditions[answer.id] = true;
              }

              return component;
            })}
          </TransitionGroup>
        </Padding>
        <RoutingRuleDestinationSelector
          id="then"
          label="THEN"
          destinations={destinations}
          onChange={handleThenChange}
          value={goto}
          data-test="select-then"
        />
      </div>
    </div>
  );
};

RoutingRule.propTypes = {
  rule: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  children: PropTypes.node,
  onDeleteRule: PropTypes.func.isRequired,
  onAddRoutingCondition: PropTypes.func.isRequired,
  onToggleConditionOption: PropTypes.func.isRequired,
  onUpdateRoutingCondition: PropTypes.func.isRequired,
  onDeleteRoutingCondition: PropTypes.func,
  onThenChange: PropTypes.func.isRequired,
  title: PropTypes.string,
  destinations: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  pagesAvailableForRouting: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
  className: PropTypes.string
};

RoutingRule.fragments = {
  RoutingRule: routingRuleFragment
};

export default RoutingRule;
