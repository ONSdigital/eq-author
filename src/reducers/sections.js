import _, {merge} from 'lodash';
import * as actions from 'actions/sections';

const ids = [];
const getIdFromName = name => {
  const newId = name.replace(/\s+/g, '-').toLowerCase();
  ids.push(newId);
  return newId;
};

const getSectionById = (sections, id) =>
  _(sections).filter({id: id}).value()[0];

const sections = (state = {sections: []}, action) => {
  switch (action.type) {
    case actions.ADD_SECTION:
      return {
        sections: [
          ...state.sections,
          {
            id: getIdFromName(action.payload.name),
            name: action.payload.name,
            questions: [],
          },
        ],
      };

    case actions.ADD_QUESTION:
      return merge(
        {},
        state.sections,
        getSectionById(
          state.sections,
          action.payload.sectionID
        ).questions.push({
          id: getIdFromName(action.payload.name),
          name: action.payload.name,
          answers: [],
        })
      );

    case actions.ADD_ANSWER:
      _(state.sections)
        .map('questions')
        .flatten()
        .filter({id: action.payload.questionID})
        .value()[0]
        .answers.push({
          id: getIdFromName(action.payload.name),
          name: action.payload.name,
        });

      return merge(
        {},
        state.sections,
        getSectionById(state.sections, action.payload.sectionID)
      );

    default:
      return state;
  }
};

export default sections;
