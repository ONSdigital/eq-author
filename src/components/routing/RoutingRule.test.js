import React from "react";
import { shallow } from "enzyme";

import RoutingRule from "./RoutingRule";

import sections from "./mockstate";

let wrapper, props;

describe("components/RoutingRule", () => {
  beforeEach(() => {
    props = {
      onAddRule: jest.fn(),
      onDeleteRule: jest.fn(),
      onThenChange: jest.fn(),
      title: "Test",
      page: sections[0].pages[0],
      sections
    };

    wrapper = shallow(
      <RoutingRule {...props}>
        <div>I am a child</div>
      </RoutingRule>
    );
  });

  it("should render children", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render empty message if no children are passed", () => {
    wrapper = shallow(<RoutingRule {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should allow deleting rule", () => {
    wrapper.find("[data-test='btn-delete']").simulate("click");
    expect(props.onDeleteRule).toHaveBeenCalled();
  });

  it("should allow change of THEN condition", () => {
    const data = { name: "hello", value: "there" };
    wrapper.find("[data-test='select-then']").simulate("change", data);
    expect(props.onThenChange).toHaveBeenLastCalledWith(data);
  });
});
