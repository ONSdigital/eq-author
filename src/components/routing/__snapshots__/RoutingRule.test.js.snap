// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RoutingRule should render children 1`] = `
<RoutingRule__RoutingRuleCanvas
  id={null}
>
  <RoutingRule__RoutingToolbar>
    <RoutingRule__Title>
      Test
    </RoutingRule__Title>
    <Toolbar__Buttons>
      <IconButtonDelete
        data-test="btn-delete"
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </IconButtonDelete>
    </Toolbar__Buttons>
  </RoutingRule__RoutingToolbar>
  <React.Fragment>
    <RoutingRule__RoutingStatement>
      <div>
        I am a child
      </div>
    </RoutingRule__RoutingStatement>
    <RoutingRuleResultSelector
      data-test="select-then"
      disabled={false}
      id="then"
      label="THEN"
      onChange={[Function]}
      sections={
        Array [
          Object {
            "id": "1",
            "pages": Array [
              Object {
                "disabled": false,
                "id": "2",
                "options": Array [
                  Object {
                    "id": "3",
                    "label": "Bibendum Cras",
                    "selected": false,
                  },
                  Object {
                    "id": "4",
                    "label": "Aenean Quam Venenatis Elit",
                    "selected": false,
                  },
                  Object {
                    "id": "5",
                    "label": "Purus Ornare Ipsum",
                    "selected": false,
                  },
                  Object {
                    "id": "6",
                    "label": "Nibh",
                    "selected": false,
                  },
                  Object {
                    "id": "7",
                    "label": "Risus Mattis Magna Bibendum",
                    "selected": false,
                  },
                  Object {
                    "id": "8",
                    "label": "Consectetur Ultricies Purus Elit Fermentum",
                    "selected": false,
                  },
                ],
                "pageType": "QuestionPage",
                "title": "1.1 Curabitur blandit tempus porttitor",
              },
              Object {
                "disabled": false,
                "id": "9",
                "options": Array [
                  Object {
                    "id": "10",
                    "label": "Yes",
                    "selected": false,
                  },
                  Object {
                    "id": "11",
                    "label": "No",
                    "selected": false,
                  },
                ],
                "pageType": "QuestionPage",
                "title": "1.2 Nulla vitae elit libero, a pharetra augue.",
              },
            ],
            "title": "1. Integer posuere erat a ante",
          },
        ]
      }
    />
  </React.Fragment>
</RoutingRule__RoutingRuleCanvas>
`;

exports[`components/RoutingRule should render empty message if no children are passed 1`] = `
<RoutingRule__RoutingRuleCanvas
  id={null}
>
  <RoutingRule__RoutingToolbar>
    <RoutingRule__Title>
      Test
    </RoutingRule__Title>
    <Toolbar__Buttons>
      <IconButtonDelete
        data-test="btn-delete"
        disabled={true}
        onClick={[Function]}
      >
        Delete
      </IconButtonDelete>
    </Toolbar__Buttons>
  </RoutingRule__RoutingToolbar>
  <RoutingRuleEmpty
    onAddRule={[Function]}
    title="No routing rules exist for this question"
  />
</RoutingRule__RoutingRuleCanvas>
`;
